File: api-service/lib/JobPosting.ts
import mongoose from "mongoose";

interface JobPostingAttributes {
  title: string;
  publicationDate: Date;
  location: string;
  company: string;
  source: string;
  url: string;
}

interface JobPostingDoc extends mongoose.Document {
  title: string;
  publicationDate: Date;
  location: string;
  company: string;
  source: string;
  url: string;
}
interface JobPostingModel extends mongoose.Model<JobPostingDoc> {}

const JobPostingSchema = new mongoose.Schema<JobPostingDoc>(
  {
    title: {
      type: String,
      required: true,
    },
    company: {
      type: String,
      required: true,
    },
    location: {
      type: String,
      required: true,
    },
    publicationDate: {
      type: mongoose.Schema.Types.Date,
      required: true,
    },
    source: {
      type: String,
      required: true,
    },
    url: {
      type: String,
      required: true,
    },
  },
  {
    toJSON: {
      transform(doc, ret) {
        ret.id = ret._id;
        delete ret._id;
      },
    },
  }
);

const JobPosting = mongoose.model<JobPostingDoc, JobPostingModel>("JobPosting", JobPostingSchema);

export { JobPosting };

File: mobile_app/lib/model/JobPosting.dart
import 'dart:convert';
import 'package:intl/intl.dart';

class JobPostingModel {
  String id;
  String title;
  String company;
  String location;
  String publicationDate;
  String source;
  String url;

  JobPostingModel(
      {required this.id,
      required this.title,
      required this.company,
      required this.location,
      required this.source,
      required this.url,
      required this.publicationDate});

  factory JobPostingModel.fromJson(Map<String, dynamic> json) {
    var publicationDate = json['publicationDate'];
    if(publicationDate != null) {
      var time = DateTime.parse(json['publicationDate']);
      publicationDate = DateFormat("EEEE, dd MMMM yyyy").format(time);
    }
    return JobPostingModel(
      id: json["id"],
      title: json["title"],
      company: json["company"],
      location: json["location"],
      source: json['source'],
      url: json['url'],
      publicationDate: publicationDate ?? "No Date"
    );
  }

  Map<String, dynamic> toJson() => {};
}
